src/indexer/main.rs

```diff
@@ -0,0 +1,144 @@
+        return Ok(metadata);
+    }
+
+    Err(ModelError::UnexpectedError.into())
+}
+
+#[tokio::main]
+pub async fn main() -> Result<()> {
+    // Initialize tracing subscriber
+    fmt().with_env_filter(EnvFilter::from_default_env()).init();
+
+    info!("Starting Indexer");
+
+    let should_terminate = Arc::new(AtomicBool::new(false));
+
+    setup_ctrlc_handler(Arc::clone(&should_terminate))?;
+
+    // Start by checking and updating the current status in the db.
+    let indexing_metadata = get_indexer_metadata().await?;
+    let router_terminator = Arc::clone(&should_terminate);
+
+    // Setup the router which allows us to query health status and operations
+    let router_handle = thread::Builder::new()
+        .name("[router]".to_owned())
+        .spawn(move || {
+            let rt = tokio::runtime::Runtime::new()?;
+
+            info!("Starting router");
+            if let Err(e) = rt.block_on(router::initialize_router(router_terminator.clone())) {
+                error!("[router] unexpected error {}", e);
+            }
+
+            info!("[router] shutting down");
+            Ok(())
+        })?;
+
+    // Start the quick indexer
+    let quick_index_config = QuickIndexConfig {
+        starting_block: indexing_metadata.indexing_starting_block_number,
+    };
+    let quick_index_terminator = Arc::clone(&should_terminate);
+
+    let quick_indexer_handle = thread::Builder::new()
+        .name("[quick_indexer]".to_owned())
+        .spawn(move || {
+            let rt = tokio::runtime::Runtime::new()?;
+
+            info!("Starting quick indexer");
+            if let Err(e) = rt.block_on(service_quick::quick_index(
+                quick_index_config,
+                quick_index_terminator,
+            )) {
+                error!("[quick_indexer] unexpected error {}", e);
+            }
+            Ok(())
+        })?;
+
+    // Start the batch indexer
+    let batch_index_config = BatchIndexConfig {
+        starting_block: indexing_metadata.indexing_starting_block_number,
+    };
+    let batch_index_terminator = Arc::clone(&should_terminate);
+
+    let batch_indexer_handle = thread::Builder::new()
+        .name("[batch_indexer]".to_owned())
+        .spawn(move || {
+            let rt = tokio::runtime::Runtime::new()?;
+
+            info!("Starting batch indexer");
+            if let Err(e) = rt.block_on(service_batch::batch_index(
+                batch_index_config,
+                batch_index_terminator,
+            )) {
+                error!("[batch_indexer] unexpected error {}", e);
+            }
+            Ok(())
+        })?;
+
+    // Wait for termination, which will join all the handles.
+    wait_for_thread_completion(vec![
+        router_handle,
+        quick_indexer_handle,
+        batch_indexer_handle,
+    ])?;
+
+    Ok(())
+}
+
+fn wait_for_thread_completion(handles: Vec<JoinHandle<Result<()>>>) -> Result<()> {
+    for handle in handles {
+        handle.join().unwrap()?;
```

Comment:
In this case I am using thread instead of tokio green threads due to the fact that using green threads will exhaust tokio thread pool and causes other process relying on tokio to hang / slow down (we had this issue in our api service). Since our services will be long running, using normal system threads will be optimal.

Reply:
> Let's avoid this kind of panics

In this case since I am actually just letting it wait until all indexer and the api decides to stop, if one of them stop or fail to resolve, I think that's considered an application error and shouldn't happen. I think it should panic in those cases.

What do you think? Cause if it gracefully shutdown without alerting us or panicking we might miss some issues.

Reply:
I still think we shoud not panic here, using unwrap() can be risky if there’s any chance that the thread handles might panic. If a thread panics the main thread will also panic, terminating your entire application unexpectedly.

I think here we should have a match with 3 arms, something like

```
match handle.join() {
            Ok(Ok(())) => { .... }
            Ok(Err(e)) => {
                error!("Thread completed with an error: {:?}", e);
            }
            Err(e) => {
                error!("Thread panicked: {:?}", e);
            }
        }
```